content/posts/2018-05-26-androids-dream.md

```diff
@@ -0,0 +1,464 @@
+UI-related components are behind interfaces as well (like in MVP, MVVM, MVI).
+It is possible to mock all network interfaces (like Retrofit declarations).
+In this case basically everything is a non-platform specific code
+with mockable external sources. That means there is a way to launch
+the whole application on the JVM, changing only a couple of things.
+Invoking actions from the UI? Sure. Receiving mocked data from the network?
+Of course. Checking UI state? Yep.
+In other words, this is a JVM-based integration testing covering
+the majority of the application. Fast, reliable and understandable.
+It even forces good abstractions as a bonus.
+
+There is a UI tests suite. The backend is mocked locally. UI interactions
+are made in a usual manner. There is a fleet of emulators either in the cloud
+or as a local infrastructure. Sometimes it works, sometimes it doesn’t.
+Runs are slow even with a state of the art sharding. Almost everything is tested,
+even the OS framework that was tested via the company which provides it.
+
+*Which one would you choose?*
+
+Personally, I would pick a good integration test covering 70% of possible
+scenarios over any UI test covering 90%. The difference is just not worth it.
+
+* Development resources spent trying to stabilize non-controlled environments
+  is a waste.
+* Anything even remotely flaky brings the opposite of confidence to developers.
+  People just press the Restart button all over again, because at this point
+  they are sure that the issue is with the environment and not the test itself.
+
+Besides, there is a technique called _Testing in Production_. This is not a joke.
+Monitoring, feature flags, A/B experiments and more — all of these are actually
+related to the in-production testing.
+There is [a great article](https://medium.com/@copyconstruct/testing-in-production-the-safe-way-18ca102d0ef1)
+about it.
+
+Recently I’ve done a technical exit interview with a former co-worker of mine.
+One of the questions I’ve asked was about the best thing in the project
+and the worst one from the technical perspective.
+
+* The best thing — BDD-style [Spek](https://github.com/spekframework/spek)
+  unit tests. It was a pure joy to write and read them.
+* The worst thing — UI tests. Their flakiness and constant battles
+  with environments wore the person down. _It was a useless waste of time._
+
+Fun fact. I’ve monitored UI tests for a couple of months with a goal to note
+all bugs caught by them. The number was zero in the end. The majority
+of risky code was caught by an extensive unit tests suite.
+
+# The Fall
+
+> I’ve seen things you people wouldn’t believe.
+> Attack ships on fire off the shoulder of Orion.
+> I watched C-beams glitter in the dark near the Tannhäuser Gate.
+> All those moments will be lost in time, like tears in rain.
+>
+> -- [*Roy Batty*](https://en.wikipedia.org/wiki/Tears_in_rain_monologue)
+
+It is always hard to let go. A lot of effort was put into writing
```

Comment:
Sooooo

Everything is alright, except that it's probably better to say something like

"We've come to a difficult decision to abandon UI tests. Grab a cup of tea and read our story"

in the beginning

Otherwise article lacks reasoning till you read it to the end. Reader might be demotivated by walls of text upfront and as an author you need to compensate that proactively if you want to keep it long (it's not bad)

Also looks like you're getting closer and closer to whitepaper-like amounts of text @sockeqwe produces hahah

Reply:
> "We've come to a difficult decision to abandon UI tests. Grab a cup of tea and read our story"

I second this 

and this 

> Also looks like you're getting closer and closer to whitepaper-like amounts of text @sockeqwe produces

😸

Reply:
Yep, that’s a good point, thanks!

About the length — this is the longest so far. Personally I’m not a fan of huge articles, but otherwise some things would not be motivated or explained enough, giving a false sense that I don’t understand what I’m talking about and just hate tests.